azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/ExtensionProfileSettings.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.azurecompute.arm.domain;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import org.jclouds.javax.annotation.Nullable;
+import org.jclouds.json.SerializedNames;
+
+import java.util.List;
+
+@AutoValue
+public abstract class ExtensionProfileSettings {
+
+   /**
+    * The fileUris reference of the extension profile settings
+    */
+   @Nullable
```

Comment:
You are enforcing presence and immutability. Remove this annotation.

Reply:
These @Nullable annotations are required to serialized the json response from the service (especially the when the value is not returned as sometimes there are not setup by the users)

Reply:
In our domain objects, implemented with Google AutoValue, the `@Nullable` annotation is taken into account inside the `AutoValue_` constructor. In these cases, a `null` value will never get there, since we're always calling it with a non-null value. The questions then re:

* If you need to be able to serialize this object and send it to the Azure API **without** this field, then we need to follow the same approach we use with the `tags`, and remove the annotation and initialize to "null or immutable", as we want to explicitly support null.
* If there is no such a need of **sending** a null value, and it's only us reading it, then we should think about how we want a Java API to look like for users:
  * Null collections is an anti-pattern, so we'd better enforce presence and immutability. This helps to build a cleaner and easier to use API.
  * In this case, as we're enforcing presence in the constructor, the annotation is not required (if I'm not wrong), as it is only used by Google AutoValue (and not for serialization/deserialization) and we're already taking care of always passing a value.

Reply:
@nacx There are two reasons why these @nulleable are valid and should stay:

1-If the nullable is removed the tests fail as there aren't any value in this case (as real response) null is appropriate.  It is not really the anti-pattern case for this, because null is not used to handle exceptions.
2- The @nulleable came from previous developed/approved code already used by our jclouds customers.  We will be having two different implementations under the same package.

Reply:
>If the nullable is removed the tests fail as there aren't any value in this case (as real response) null is appropriate. It is not really the anti-pattern case for this because null is not used to handle exceptions.

I've built your branch and removed the annotation and tests don't fail.
A jclouds API is **not** an exact mapping to existing REST APIs. We go one step further to provide a meaningful experience when consuming an API from Java. We take care of properly mapping response error codes to appropriate exceptions, building good transfer objects, etc. Part of this work is helping users use clean APIs (from a Java perspective), and that means reducing the changes where users can make mistakes. Having `nullable` lists is an anti-pattern that should be avoided, as empty lists are a perfect substitute and make much more sense.

My point is that **your code is already enforcing that**. The code in the `create` method already enforces an empty list when a `null` value (or absent value) cames in the response. With the code in this factory method, **there is no way** an instance of an `ExtensionProfileSettings` object can have a null value in this field, thus, the `@Nullable` annotation is simply not true, as users will never get a `null` value in these objects. Please, remove it and apply this pattern in all the mentioned domain objects.

As I said, the rule of thumb for a consistent implementation and declaration is:

* If your constructor/factory methods enforce presence (as we do in most cases), the `@Nullable` annotation does not apply and should not be there.
* If there is a good reason to not enforce presence in the constructor/factory method (like in tags), then let's not enforce presence and annotate the field accordingly.

Here is a diff you can apply to verify that there are no serialization/deserialization or other issues if you remove the annotation, even if the response does not contain a value for that field. Tests keep passing.

```diff
diff --git a/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/ExtensionProfileSettings.java b/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/ExtensionProfileSettings.java
index 0d89fda3e..61d50d910 100644
--- a/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/ExtensionProfileSettings.java
+++ b/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/ExtensionProfileSettings.java
@@ -16,12 +16,12 @@
  */
 package org.jclouds.azurecompute.arm.domain;
 
-import com.google.auto.value.AutoValue;
-import com.google.common.collect.ImmutableList;
-import org.jclouds.javax.annotation.Nullable;
+import java.util.List;
+
 import org.jclouds.json.SerializedNames;
 
-import java.util.List;
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
 
 @AutoValue
 public abstract class ExtensionProfileSettings {
@@ -29,7 +29,6 @@ public abstract class ExtensionProfileSettings {
    /**
     * The fileUris reference of the extension profile settings
     */
-   @Nullable
    public abstract List<String> fileUris();
 
    /**
diff --git a/azurecompute-arm/src/test/resources/createvirtualmachinescalesetresponse.json b/azurecompute-arm/src/test/resources/createvirtualmachinescalesetresponse.json
index ff2971d40..4b118c50a 100644
--- a/azurecompute-arm/src/test/resources/createvirtualmachinescalesetresponse.json
+++ b/azurecompute-arm/src/test/resources/createvirtualmachinescalesetresponse.json
@@ -42,7 +42,7 @@
               "type": "CustomScriptExtension",
               "typeHandlerVersion": "1.1",
               "autoUpgradeMinorVersion": false,
-              "settings": {"fileUris":["https://mystorage1.blob.core.windows.net/winvmextekfacnt/SampleCmd_1.cmd"],"commandToExecute":"SampleCmd_1.cmd"}
+              "settings": {"commandToExecute":"SampleCmd_1.cmd"}
             },
             "name": "extensionName"
           }
diff --git a/azurecompute-arm/src/test/resources/virtualmachinescaleset.json b/azurecompute-arm/src/test/resources/virtualmachinescaleset.json
index c99eb39a5..720f71ff8 100644
--- a/azurecompute-arm/src/test/resources/virtualmachinescaleset.json
+++ b/azurecompute-arm/src/test/resources/virtualmachinescaleset.json
@@ -42,7 +42,7 @@
               "type": "CustomScriptExtension",
               "typeHandlerVersion": "1.1",
               "autoUpgradeMinorVersion": false,
-              "settings": {"fileUris":["https://mystorage1.blob.core.windows.net/winvmextekfacnt/SampleCmd_1.cmd"],"commandToExecute":"SampleCmd_1.cmd"}
+              "settings": {"commandToExecute":"SampleCmd_1.cmd"}
             },
             "name": "extensionName"
           }
```